### PR TITLE
Adds lazy reader support for long strings

### PR DESCRIPTION
**Pull request by zslayton**
Builds on outstanding PRs #612, #613, #614, #616, #617, #619, #620, #621, #622, #623, #627, #628 and #629.

Adds lazy reader support for long strings.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/lazy/text/buffer.rs [hunk 1]
```diff
@@ -30,20 +30,25 @@ use crate::{IonError, IonResult, IonType, TimestampPrecision};
 
 impl<'a> Debug for TextBufferView<'a> {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        const CHARS_TO_SHOW: usize = 64;
```
Comment (zslayton) [inline src/lazy/text/buffer.rs:33]: 🗺️ Changes in this method are just introducing a named constant and bumping the number of characters printed from `32` to `64`.

File: src/lazy/text/buffer.rs [hunk 2]
```diff
@@ -1001,10 +1006,41 @@ impl<'data> TextBufferView<'data> {
         Self::match_text_until_unescaped(self, b'\"')
     }
 
-    fn match_long_string(self) -> IonParseResult<'data, MatchedString> {
-        // TODO: implement long string matching
-        //       The `fail` parser is a nom builtin that never matches.
-        fail(self)
+    /// Matches a long string comprised of any number of `'''`-enclosed segments interleaved
+    /// with optional comments and whitespace.
+    pub fn match_long_string(self) -> IonParseResult<'data, MatchedString> {
```
Comment (zslayton) [inline src/lazy/text/buffer.rs:1011]: 🗺️ I recommend going to look at the `MatchedString` enum before reading this implementation. It explains why we're tracking the number of segments and the presence of escaped characters.

File: src/lazy/text/buffer.rs [hunk 3]
```diff
@@ -1597,11 +1671,11 @@ mod tests {
     }
 
     impl MatchTest {
-        /// Takes an `input` string and appends a trailing space to it, guaranteeing that the
+        /// Takes an `input` string and appends a trailing value to it, guaranteeing that the
         /// contents of the input are considered a complete token.
         fn new(input: &str) -> Self {
             MatchTest {
-                input: format!("{input} "), // add trailing space
+                input: format!("{input}\n0"), // add whitespace and a trailing value
```
Comment (zslayton) [inline src/lazy/text/buffer.rs:1678]: 🗺️ Long string syntax is very greedy. It's impossible to declare a long string "complete" until you've run into something that _isn't_ a long string. Later on I'll add support for declaring that the input stream is done, but for the moment this workaround allows me to use the same test infrastructure on long strings.

**Review by zslayton — COMMENTED** [review]
🗺️ PR tour

File: src/lazy/text/buffer.rs [hunk 4]
```diff
@@ -1889,6 +1963,15 @@ mod tests {
             r#"
             "this has an escaped quote \" right in the middle"
             "#,
+            r#" '''hi''' "#,
```
Comment (popematt) [inline src/lazy/text/buffer.rs:1966]: Add case for empty string.
```suggestion
            r#" '''hi''' "#,
            r#" ''''''  '''''' "#,
```

File: src/lazy/text/matched.rs [hunk 5]
```diff
@@ -286,15 +287,15 @@ pub(crate) enum MatchedString {
...
-    Long(MatchedLongString),
-}
-
-#[derive(Clone, Copy, Debug, PartialEq)]
-pub(crate) struct MatchedLongString {
-    // TODO: Decide what (if anything) to store here.
-    //       Storing any collection of bytes or ranges means that this type cannot implement Copy,
-    //       which in turn means MatchedValue and EncodedTextValue also cannot implement Copy.
-    //       We probably also don't want to heap allocate just to match the long string.
+    Long,
```
Comment (popematt) [inline src/lazy/text/matched.rs:290]: Just a thought—would it be useful/advantageous to have an `Empty` variant for an empty string?

**Review by popematt — APPROVED** [review]

File: src/lazy/text/matched.rs (hunk 5 above)
Comment (zslayton) [reply src/lazy/text/matched.rs:290]: An empty string would be considered a `LongSingleSegmentWithoutEscapes`, yielding a reference to an empty slice of the input buffer. It should be close to optimal.

**Pull request merged**